bundles/org.openhab.ui.basic/src/main/java/org/openhab/ui/basic/internal/render/InputRenderer.java

```diff
@@ -58,38 +69,127 @@ public boolean canRender(Widget w) {
 
     @Override
     public EList<Widget> renderWidget(Widget w, StringBuilder sb, String sitemap) throws RenderException {
+        Input input = (Input) w;
+
         String snippet = getSnippet("input");
 
         snippet = preprocessSnippet(snippet, w);
 
-        String dataState = getValue(w);
-        State state = itemUIRegistry.getState(w);
-        if (state == null || state instanceof UnDefType) {
-            snippet = snippet.replace("%undef_state%", dataState);
-            snippet = snippet.replace("%data_state%", "");
-        } else {
-            snippet = snippet.replace("%undef_state%", "");
-            snippet = snippet.replace("%data_state%", dataState);
-        }
-
         Item item = null;
         try {
             item = itemUIRegistry.getItem(w.getItem());
         } catch (ItemNotFoundException e) {
             logger.debug("Failed to retrieve item during widget rendering: {}", e.getMessage());
         }
         String dataType;
-        if (item != null && item.getAcceptedDataTypes().stream().anyMatch(o -> Number.class.isAssignableFrom(o))) {
-            dataType = "Number";
+        if (item != null && item.getAcceptedCommandTypes().stream()
+                .anyMatch(o -> (!StringType.class.isAssignableFrom(o) && Number.class.isAssignableFrom(o)))) {
+            dataType = "number";
+        } else if (item != null && item.getAcceptedCommandTypes().stream()
+                .anyMatch(o -> (!StringType.class.isAssignableFrom(o) && DateTimeType.class.isAssignableFrom(o)))) {
+            dataType = "datetime";
         } else {
-            dataType = "Text";
+            dataType = "text";
         }
         snippet = snippet.replace("%data_type%", dataType);
 
+        String inputHint = input.getInputHint();
+        snippet = snippet.replace("%input_hint%", inputHint == null ? "" : inputHint);
+
+        String inputType = "text";
+        String inputPattern = "";
+        if ("number".equals(inputHint)) {
+            inputType = "number";
+            inputPattern = "pattern=\"[0-9]*([\\.|,][0-9]*)?\"";
+        } else if ("date".equals(inputHint)) {
+            inputType = "date";
+            inputPattern = "pattern=\"[0-9]{4}-[0-9]d{2}-[0-9]d{2}\"";
+        } else if ("time".equals(inputHint)) {
+            inputType = "time";
+            inputPattern = "pattern=\"[0-9]{2}:[0-9]{2}(:[0-9]{2})?\"";
+        } else if ("datetime".equals(inputHint)) {
+            inputType = "datetime-local";
+            inputPattern = "pattern=\"[0-9]{4}-[0-9]d{2}-[0-9]d{2} [0-9]{2}:[0-9]{2}(:[0-9]{2})?\"";
+        }
+        snippet = snippet.replace("%input_type%", inputType);
+        snippet = snippet.replace("%input_pattern%", inputPattern);
+
+        String displayState = getValue(w);
+        State state = itemUIRegistry.getState(w);
+
+        String undefState = "-";
```

Comment:
That means `"%undef_state%"` will now be replaced by "-" when the item is a number or string item with a state ?
It was before set to "" which made sense.

Reply:
In principle, undef_state can be set once and never touched anymore. It will only show when the data value is empty. So in almost all cases, it would be fine to have "-". But indeed, we ran into the issue that for text values, an empty string is acceptable, and would show the undef_value. I forgot about it here. So went back to having default "" and updating undef_state from the SSE return value in js.